llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp

```diff
@@ -2580,6 +2586,37 @@ SDValue NVPTXTargetLowering::LowerFROUND64(SDValue Op,
   return DAG.getNode(ISD::SELECT, SL, VT, IsLarge, A, RoundedA);
 }
 
+SDValue NVPTXTargetLowering::LowerINT_TO_FP(SDValue Op,
+                                            SelectionDAG &DAG) const {
+  // sm_90 has instructions for bf16 conversions, sm_80 only has f32 -> bf16.
+  if (Op.getValueType() == MVT::bf16 &&
+      (STI.getSmVersion() < 90 || STI.getPTXVersion() < 78)) {
+    SDLoc Loc(Op);
+    return DAG.getNode(
+        ISD::FP_ROUND, Loc, MVT::bf16,
```

Comment:
LLVM also has `BF16_TO_FP` op, which may be a better fit than generic FP_ROUND. I'm not sure if it would buy us anything useful here. Looks like LLVM would try to lower it to integer extend to i32 and `shl 16`. It may give LLVM a bit more freedom, though I suspect PTX own conversion bf16->fp32 would expand into about the same SASS in the end.

Reply:
`BF16_TO_FP` is a bit of an odd beast that's only used during legalization. We can't use it here because it's not a legal op. I would be really surprised if the SASS is different.